### ISSUE TITLE
ENH: sorted numpy array of datetime64 as pandas.DataFrame.index without copy-conversion

### ISSUE THREAD

**Issue by max0x7ba** [pandas-dev/pandas#45358]
#### Is your feature request related to a problem?
Presently, using any `datetime64[unit]` `numpy` array as a `pandas.DataFrame` index results in copy-conversion to an array of `pandas.Timestamp`.

In my particular use case, I would like to construct a `pandas.DataFrame` from existing `numpy` arrays as a view using an existing sorted `datetime64[ms]` `numpy` array as index view, with no copy-conversion. That works for `numpy.float64` `numpy` arrays which don't get copied, but my `datetime64[ms]` `numpy` array gets copy-converted to an array of `pandas.Timestamp` when used as an index.

I would like to use a sorted `datetime64[unit]` `numpy` array as a `pandas.DataFrame` index as is, with no copy-conversion. 

#### Describe the solution you'd like
One solution can be a new `pandas` index type that uses an array of sorted `datetime64[unit]` as is, instead of converting it to an index type with `pandas.Timestamp` elements. 

Another solution is a new `pandas` index type that uses any sorted array as index as is, when `copy=False`.

These indexes can use binary search for sub-linear complexity index lookup. Binary search lookup may not be as fast as a hashtable lookup, but the former doesn't require constructing any extra data structures. Any sorted array should be usable as an index with binary lookup and no copy-conversion.

#### API breaking implications
No breaking of existing APIs should be necessary.

These can be new index types implementing existing index APIs, which requires no changes to any existing API user code. I am not familiar with `pandas` implementation, though.

#### Describe alternatives you've considered
I could probably work-around this issue by using milliseconds-since-epoch integer timestamps instead of stronger-typed `datetime64[ms]` converted to `pandas.Timestamp` as index (if that's possible at all to use a plain sorted integer array as an index for `pandas.DataFrame` without copying and memory allocation (e.g. binary search the index)), but that defeats the index/row type-based assertions, so that my row indexing mistakes when index is used where timestamp is required can no longer be detected by an IDE or run-time typechecks. I am not sure whether any sorted array can be used as a `pandas` index with no copy-conversion.

#### Additional context
A test case:
```
def test_numpy_view():
    import numpy as np
    import pandas as pd

    start_date = np.datetime64("2022-01-01")
    step = np.timedelta64(1, "m")
    steps = 3
    times = np.arange(start_date, start_date + step * steps, step).astype("datetime64[ms]")

    data = np.zeros(steps)

    s = pd.Series(data=data, index=times, copy=False)

    assert data[0] == s.iloc[0]
    data[0] -= 1
    assert data[0] == s.iloc[0] # Succeeds as expected, s is a view of data.

    assert times[0] == s.index[0]
    times[0] -= step
    assert times[0] == s.index[0] # Fails because s.index is not a view of times.
```

**Comment by jreback** [comment]
this is adding quite a bit of complexity - what is the purpose of this? what is gained by doing this?

note that the indexes don't actually convert to Timestamps rather they are stored as numpy i8 under the hood

**Comment by max0x7ba** [comment]
The purpose is to represent existing arrays in shared memory as `pandas.DataFrame`, along with the index represented as 64-bit timestamp, without any data copy. 

I.e. a publisher maintains arrays of arithmetic types along with an array of timestamps. A subscriber maps those arrays as numpy arrays and that works as expected.

I would like these arrays to be columns of `pandas.DataFrame` with the index being one of those  `datetime64[ms]` arrays. 

The array values don't change while a `pandas.DataFrame` refers to them, it just refers to existing immutable data.

I cannot afford copying data but that convertion of 64-bit timestamps to `pandas.Timestamp` defeats the idea.

Does that make sense to you?

**Comment by jreback** [comment]
you missed the point

these ARE already numpy arrays

**Comment by max0x7ba** [comment]
> you missed the point
> 
> these ARE already numpy arrays

I am not sure I follow you. Could you rephrase?

**Comment by mroeschke** [comment]
Conversion to a `pd.Timestamp` only happens when "boxing" needs to occur (e.g. iloc) not during storage i.e. the underlying data _is_ a numpy array

```
In [1]: DatetimeIndex(range(5)).values
Out[1]:
array(['1970-01-01T00:00:00.000000000', '1970-01-01T00:00:00.000000001',
       '1970-01-01T00:00:00.000000002', '1970-01-01T00:00:00.000000003',
       '1970-01-01T00:00:00.000000004'], dtype='datetime64[ns]')
```

**Comment by max0x7ba** [comment]
> Conversion to a `pd.Timestamp` only happens when "boxing" needs to occur (e.g. iloc) not during storage i.e. the underlying data _is_ a numpy array

That would be great. 

My goal is to wrap numpy arrays with a `pd.DataFrame` interface without copying the numpy arrays. My original test case demonstrates that the data is not copied, as expected. But the index is copied, see the last 3 lines of the test case.

Is my test case wrong? 

If not, how can I make `pd.DataFrame` use my original numpy datetime64 array as index without copy?

**Comment by max0x7ba** [comment]
> you missed the point
> 
> these ARE already numpy arrays

I understand what you mean now. 

The index may be a numpy array, but the last 3 lines of my test case demonstrate that the index isn't the original array but rather a copy. My goal is to avoid copying the index.

**Comment by jreback** [comment]
it's copied because u r using a non ns precision

**Comment by max0x7ba** [comment]
> it's copied because u r using a non ns precision

I changed my index array element type to `datetime64[ns]` and that fixed the test case failure on the last line. That's exactly what I need, thank you very much.

**Comment by jreback** [comment]
welcome
